### ISSUE TITLE
More colum filter options in browser

### ISSUE THREAD

**Issue by tlhackque** [sqlitebrowser/sqlitebrowser#1463]
Describe the new feature
--------------------------
It would be nice to have some options besides 'like' when filtering in the browser.

Similar to Excel's "blanks", "non-blanks" in Autofiltered columns.  (I'm not a big fan of "top 10")

I'd add "NULL" & "not NULL" as you can't type anything to represent them, and perhaps "Blob" and "not Blob" for the same reason.  These are helpful when looking for anomalies.

It's possible to get at NULL values for a single column by sorting & scanning visually, but that doesn't scale.

It's possible to drop into SQL to do this, but that result table can't be filtered (except by writing more SQL).  Hmmm - would it be possible to have the browser take the results of my query and load them into the filterable window?  e.g. create a temporary table from the results of a select in the "execute SQL' tab, and put that into your nice filterable table.    Seems like that would be straightforward for viewing (my primary use case), but harder to send modifications back to the real tables.

It also would be useful to add a simple JOIN column (e.g. viewing orders, would be nice to add 'join order.customerid = customers.id')

Additional info
---------------
Please answer these questions before submitting your feature request.

#### Is your feature request related to an issue? Please include the issue number.
No

#### Does this feature exist in another product or project? Please provide a link.
Excel.

#### Do you have a screenshot? Please add screenshots to help explain your idea.
Nope.  Just run excel autofilter on any spreadsheet and look at the dropdowns.  [All] [Blanks] [Non-Blanks] [Custom]

**Comment by justinclift** [comment]
Some extra possibilities are already available:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/wiki/Using-the-Filters

That being said, in theory you should be able to save the results from the Execute SQL tab as a View.  From there, you should be able to apply filtering to that.

It's not exactly a temporary table, but might be "good enough" for the moment?

**Comment by tlhackque** [comment]
Those aren't exactly obvious on first approach - how about a help->'filters' window with that information?  And / or a right-click in the filter cell that offers help?  I did try both early on.

Still doesn't give me '!Null' (or any variation that I tried'). or =isNULL

> theory you should be able to save the results from the Execute SQL tab as a View

Try saving as a view with a read-only database :-)

[Hint - it's grayed-out].  I'm not going to try exporting 100K records to a CSV (which isn't.)

Which seems to indicate that to do what I want, you'd need a temporary DB in this case.

**Comment by justinclift** [comment]
> how about a help->'filters' window with that information?
> And / or a right-click in the filter cell that offers help?

Good thinking.  We've just been doing the wiki thing so far as that seemed good enough.

I wonder if clickable links work in tooptip text?  That might be a useful way to add "Further info: _url_" type of things too.

> Try saving as a view with a read-only database :-)

Oops.  Yeah.  Good point. :wink:

Temp database does sound like it's worth trying as a workaround for now then.

**Comment by tlhackque** [comment]
> "Further info: url" 

I'd rather not see this sort of help require a browser; it should be built-in.  It's pretty small, though big for a tooltip.  A floating (non-modal) dialog box would work.

Browsers are slow - and if you mean to assume on-line - I get a lot done at 30,000 ft.  Shouldn't need an internet connection for a couple of paragraphs.    Plus, they'll get translated if they're in the product, and less likely to be (or timely) if on a wiki page.

One easy clue to add would be to have the "filter" cue be "filter expression" - which at least will cause wonder about what the expression syntax might be...

**Comment by justinclift** [comment]
Good point.  We probably need to think about doing a "real" help system then. :smile:

**Comment by mgrojo** [comment]
Have you checked the "What's this" information for the filter box? It is not extensive, but it should give hints about the possible filter options. In my latest commits I've tried to add useful tips for the more hidden features, but I always have the feeling, that users are not used to read these "What's this" tips :smile:

**Comment by tlhackque** [comment]
I did not notice that feature.  Sorry :-( 

It's good information.  It could be more prominent.

Suggest elevating it to an icon on the toolbar and making the (same) context help available on a right click in a field.  Keeping the help near the task usually wins.

**Comment by n-arakawa** [comment]
> Still doesn't give me '!Null' (or any variation that I tried'). or =isNULL
In addition to this, I hope to have 'not like' filter.

**Comment by mgrojo** [comment]
I added a commit for improving the visibility of the different filter options. See my comment in bbe2e33ea4f1f22c85d133bdc8af20ea986eea2f for explanatioon. It should be already in the nightly.

**Comment by tlhackque** [comment]
Nicely done!

**Comment by mgrojo** [comment]
Recollecting, so far we have implemented:
- [X] Visibility and helper menus of existing filters
The pending enhancements mentioned in this issue are these kind of filters:
- [ ] Is BLOB, Not is BLOB
- [ ] Not like 'X'

**Comment by tlhackque** [comment]
> The pending enhancements mentioned in this issue are these kind of filters:
...
And (from .0):

> It also would be useful to add a simple JOIN column (e.g. viewing orders, would be nice to add 'join order.customerid = customers.id')

It's not a filter (the idea was "add column by creating an implicit temp view") - Should another issue be opened so it's not lost?  

And (.5, justin) might also want to be split out:

> We probably need to think about doing a "real" help system then.

Thanks again for all the work.

**Comment by mgrojo** [comment]
@tlhackque:
> >    It also would be useful to add a simple JOIN column (e.g. viewing orders, would be nice to add 'join order.customerid = customers.id')

> It's not a filter (the idea was "add column by creating an implicit temp view") - Should another issue be opened so it's not lost?

Yes, please, could you open a new issue and explain the idea in more detail?

**Comment by dt192** [comment]
I'd like to see SQL query as a filter type.

Currently creating a view, then setting up the update triggers and then enabling editing is more work than it's worth for ad-hoc data edits.

Example would be, I set an sql filter on `file_id` column containing `SELECT DISTINCT file_id FROM my_table WHERE content = 'something specific'` this would show me all substrings of each file that contain the substring 'something specific'.

This would also mean I could use data from a different table as a filter.
